db/snapshot/backup_task.cc

```diff
@@ -113,6 +114,10 @@ future<> backup_task_impl::do_backup() {
     }
 
     std::exception_ptr ex;
+    // FIXME: this should be configurable, or derived dynamically from
+    // the concurrency level of S3 uploads
+    constexpr size_t max_concurrency = sstables::num_component_types * 2;
+    auto concurrency_sem = semaphore(max_concurrency);
```

Comment:
There's `sstables::directory_semaphore` already that limits processing of sstables directory. It would be nice not to add more concurrency control on top of the existing one

Reply:
But they control different resources.
The sstable directory semaphore is about listing in the local filesysten while this one is about throttling sstable upload over the network

Reply:
changed algorithm to use bytes-in-flight in v2 ([06da3f3](https://github.com/scylladb/scylladb/commit/06da3f30f0c2da5e684fbff7b517e97d70d70559))

Reply:
Why do we want to limit the bytes-in-flight? To keep memory from growing or to allow for queue prioritization? If the former, then there's s3_clients_memory semaphore out there already, if the latter, then sstable dir sem should work just fine (all the more so it's not just _listing_, but it's all about load parallelism whatever it means)

Reply:
It's the latter (queue prioritization), just as @avikivity [observed](https://github.com/scylladb/scylladb/pull/23241#discussion_r1999088298), we want to limit parallelism of bytes in flight, but if the s3_clients_memory semaphore already does that, then you have a valid point and we can just reuse the sstable dir sem to limit count for queue prioritization, and let the s3_clients_memory semaphore do limit bytes-in-flight.

Reply:
In v3 ([7ab3391](https://github.com/scylladb/scylladb/commit/7ab3391f15e68adfff6187385be438b96bb3d57e)): using the sstable directory semaphore.